File: test/python/fixtures/generate_v1_fixtures.py
"""Generate the v1 (schema_version=1) golden-fixture matrix for the unified envelope.

Run once to (re)mint the checked-in v1 fixtures:

    python test/python/fixtures/generate_v1_fixtures.py

The matrix is {bart, bcf} x {numeric, categorical} x {no-rfx, rfx} = 8 models,
covering the format-relevant axes: forest naming (BART vs BCF), the covariate
preprocessor (identity vs categorical encoding -> the cross-platform-portable
axis), and random effects (which add the random_effects subfolder).

Fixtures only need to be structurally complete and loadable, not statistically
meaningful, so the ensembles are tiny (5 trees, 2 MCMC) to keep the checked-in
files small (a few KB each). The legacy v0 fixtures (bart_mcmc.json /
bcf_mcmc.json) are kept separately and exercise the v0 -> v1 migration path.
"""

import json
from pathlib import Path

import numpy as np
import pandas as pd

from stochtree import BARTModel, BCFModel

OUT = Path(__file__).parent

_NUM_MCMC = 2
_SMALL = {"num_trees": 5}
N = 200
P = 4  # numeric covariate columns (categorical models add one "cat" column)

# Superseded ad-hoc fixtures from the first cut (replaced by the matrix below).
_SUPERSEDED = ("bart_mcmc_v1.json", "bart_rfx_v1.json", "bcf_mcmc_v1.json")


def _make_covariates(rng, n, categorical):
    """Return (X, x0) where x0 is a numeric column used to build the outcome."""
    X_num = rng.uniform(size=(n, P))
    if not categorical:
        return X_num, X_num[:, 0]
    X = pd.DataFrame(X_num, columns=[f"x{i}" for i in range(P)])
    X["cat"] = pd.Categorical(rng.choice(["a", "b", "c"], size=n))
    return X, X_num[:, 0]


def _bart(seed, *, rfx, categorical):
    rng = np.random.default_rng(seed)
    X, x0 = _make_covariates(rng, N, categorical)
    y = x0 + rng.normal(scale=0.5, size=N)
    kw = dict(num_gfr=0, num_burnin=0, num_mcmc=_NUM_MCMC, mean_forest_params=_SMALL)
    if rfx:
        g = rng.integers(0, 3, size=N)
        y = y + g
        kw["rfx_group_ids_train"] = g
        kw["rfx_basis_train"] = np.ones((N, 1))
    m = BARTModel()
    m.sample(X_train=X, y_train=y, **kw)
    return m.to_json()


def _bcf(seed, *, rfx, categorical):
    rng = np.random.default_rng(seed)
    X, x0 = _make_covariates(rng, N, categorical)
    pi = 0.25 + 0.5 * x0
    Z = rng.binomial(1, pi).astype(float)
    y = pi * 5 + Z * x0 * 2 + rng.normal(size=N)
    kw = dict(
        num_gfr=0,
        num_burnin=0,
        num_mcmc=_NUM_MCMC,
        prognostic_forest_params=_SMALL,
        treatment_effect_forest_params=_SMALL,
    )
    if rfx:
        g = rng.integers(0, 3, size=N)
        y = y + g
        kw["rfx_group_ids_train"] = g
        kw["rfx_basis_train"] = np.ones((N, 1))
    m = BCFModel()
    m.sample(X_train=X, Z_train=Z, y_train=y, propensity_train=pi, **kw)
    return m.to_json()


def main():
    seed = 110
    for model, fn in (("bart", _bart), ("bcf", _bcf)):
        for categorical in (False, True):
            for rfx in (False, True):
                kind = "categorical" if categorical else "numeric"
                suffix = "_rfx" if rfx else ""
                name = f"{model}_{kind}{suffix}_v1.json"
                js = fn(seed, rfx=rfx, categorical=categorical)
                seed += 1
                payload = json.loads(js)
                assert payload["schema_version"] == 1, f"{name}: expected schema_version=1"
                (OUT / name).write_text(js)
                print(f"wrote {name} (schema_version={payload['schema_version']})")

    for old in _SUPERSEDED:
        p = OUT / old
        if p.exists():
            p.unlink()
            print(f"removed superseded {old}")


if __name__ == "__main__":
    main()

File: test/R/testthat/fixtures/generate_v1_fixtures.R
# Generate the v1 (schema_version=1) golden-fixture matrix for the unified envelope.
#
# Run once to (re)mint the checked-in v1 fixtures:
#   NOT_CRAN=true Rscript -e "devtools::load_all('.'); \
#     source('test/R/testthat/fixtures/generate_v1_fixtures.R')"
#
# The matrix is {bart, bcf} x {numeric, categorical} x {no-rfx, rfx} = 8 models,
# covering the format-relevant axes: forest naming (BART vs BCF), the covariate
# preprocessor (identity vs categorical encoding -> the cross-platform-portable
# axis), and random effects (which add the random_effects subfolder, including
# the relocated rfx_unique_group_ids).
#
# Fixtures only need to be structurally complete and loadable, not statistically
# meaningful, so the ensembles are tiny (5 trees, 2 MCMC) to keep the checked-in
# files small. The legacy v0 fixtures (bart_mcmc.json / bcf_mcmc.json) are kept
# separately and exercise the v0 -> v1 migration path.

out_dir <- "test/R/testthat/fixtures"

NUM_MCMC <- 2
SMALL <- list(num_trees = 5)
N <- 200
P <- 4 # numeric covariate columns (categorical models add one "cat" column)

# Superseded ad-hoc fixtures from the first cut (replaced by the matrix below).
SUPERSEDED <- c("bart_mcmc_v1.json", "bart_rfx_v1.json", "bcf_mcmc_v1.json")

make_covariates <- function(categorical) {
  X_num <- matrix(runif(N * P), ncol = P)
  if (!categorical) {
    return(list(X = X_num, x0 = X_num[, 1]))
  }
  X <- data.frame(X_num)
  X$cat <- factor(sample(c("a", "b", "c"), N, replace = TRUE))
  list(X = X, x0 = X_num[, 1])
}

bart_fixture <- function(seed, rfx, categorical) {
  set.seed(seed)
  cv <- make_covariates(categorical)
  y <- cv$x0 + rnorm(N, 0, 0.5)
  args <- list(
    X_train = cv$X,
    num_gfr = 0,
    num_burnin = 0,
    num_mcmc = NUM_MCMC,
    mean_forest_params = SMALL
  )
  if (rfx) {
    g <- sample(0:2, N, replace = TRUE)
    y <- y + g
    args$rfx_group_ids_train <- g
    args$rfx_basis_train <- matrix(1, nrow = N, ncol = 1)
  }
  args$y_train <- y
  saveBARTModelToJsonString(do.call(bart, args))
}

bcf_fixture <- function(seed, rfx, categorical) {
  set.seed(seed)
  cv <- make_covariates(categorical)
  pi_x <- 0.25 + 0.5 * cv$x0
  Z <- rbinom(N, 1, pi_x)
  y <- pi_x * 5 + Z * cv$x0 * 2 + rnorm(N)
  args <- list(
    X_train = cv$X,
    Z_train = Z,
    propensity_train = pi_x,
    num_gfr = 0,
    num_burnin = 0,
    num_mcmc = NUM_MCMC,
    prognostic_forest_params = SMALL,
    treatment_effect_forest_params = SMALL
  )
  if (rfx) {
    g <- sample(0:2, N, replace = TRUE)
    y <- y + g
    args$rfx_group_ids_train <- g
    args$rfx_basis_train <- matrix(1, nrow = N, ncol = 1)
  }
  args$y_train <- y
  saveBCFModelToJsonString(do.call(bcf, args))
}

seed <- 110
for (model in c("bart", "bcf")) {
  fn <- if (model == "bart") bart_fixture else bcf_fixture
  for (categorical in c(FALSE, TRUE)) {
    for (rfx in c(FALSE, TRUE)) {
      kind <- if (categorical) "categorical" else "numeric"
      suffix <- if (rfx) "_rfx" else ""
      name <- paste0(model, "_", kind, suffix, "_v1.json")
      js <- fn(seed, rfx, categorical)
      seed <- seed + 1
      ver <- jsonlite::fromJSON(js, simplifyVector = FALSE)$schema_version
      stopifnot(ver == 1)
      writeLines(js, file.path(out_dir, name))
      cat("wrote", name, "(schema_version =", ver, ")\n")
    }
  }
}

for (old in SUPERSEDED) {
  p <- file.path(out_dir, old)
  if (file.exists(p)) {
    file.remove(p)
    cat("removed superseded", old, "\n")
  }
}
